### PR TITLE
Add workflow to build and publish backend images

### PR DESCRIPTION
**Pull request by evan-hataishi**
**Which issue is resolved by this Pull Request:** 
Resolves https://github.com/kubeflow/kfp-tekton/issues/415

**Description of your changes:**
Add new workflow to build and publish apiserver, persistenceagent,  metadata-writer, and scheduledworkflow docker images to dockerhub. This action should execute whenever new code is pushed to (and only to) master. As it is currently written, images will be pushed to aipipeline and tagged with the short version of latest commit sha.

To test, I set `DOCKER_REGISTRY: evanhataishi` and added this branch `docker-workflow` to branches in `docker-publish.yml`. The action runs are shown under my [repo](https://github.com/evan-hataishi/kfp-tekton/actions). The pushed images can be found on my personal dockerhub [account](https://hub.docker.com/u/evanhataishi).

**Environment tested:**

* Python Version (use `python --version`):
* Tekton Version (use `tkn version`):
* Kubernetes Version (use `kubectl version`):
* OS (e.g. from `/etc/os-release`):

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

### CODE REVIEW

**Comment by Tomcli** [comment]
Thanks @evan-hataishi, can you describe how to add the secrets and env for the docker credentials? We will create a bot account for pushing these images.

File: .github/workflows/docker-publish.yml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      -
+        run: echo "TAG=`echo ${GITHUB_SHA} | cut -c1-7`" >> $GITHUB_ENV
+      -
+        name: Build and Push Image
+        uses: docker/build-push-action@v2
+        with:
+          push: true
+          file: backend/Dockerfile
+          tags: ${{ env.DOCKER_REGISTRY }}/api-server:${{ env.TAG }}
```
Comment (Tomcli) [inline .github/workflows/docker-publish.yml:30]: can we have another push to push this image to the `latest` tag? We will only be pushing images for commits that made to the master branch.

**Comment by evan-hataishi** [comment]
@Tomcli You beat me to it :). I added a comment in the original issue.

I was trying to look for ways to shorten everything. It looks like there isn't much support yet for reusing steps/jobs. Each image build/push requires the same setup. Since we want to run them all in parallel, they need to be in separate jobs (tasks only run sequentially 😢). That's why there's copy/pasted code for each job. One possibility is to use a [composite run](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action). I'm looking into this, but you may just want to just get this initial version merged in first.

**Comment by evan-hataishi** [comment]
/hold

Encrypted secrets need to be set on this repository before this is merged (described in https://github.com/kubeflow/kfp-tekton/issues/415)

File: .github/workflows/docker-publish.yml (hunk 1 above)
Comment (evan-hataishi) [reply .github/workflows/docker-publish.yml:30]: Thanks @Tomcli. Do we still want both or just latest?

Comment (Tomcli) [reply .github/workflows/docker-publish.yml:30]: you can set tag to latest for now, I think dockerhub added some limit on the number of tag we can have on free tier. We will add the commit tag later on after we figure out a better clean up workflow.

**Comment by evan-hataishi** [comment]
Thanks @Tomcli. Unfortunately, the composite run doesn't support using actions within actions (documented in this [issue](https://github.com/actions/runner/issues/646)). This seems to be a popular use case, but there hasn't been any good workaround yet. It looks like this won't get any shorter unless we decide not to use the pre-built docker actions.

Anyways, this should be good to go after the secrets are added to the repository.

**Comment by evan-hataishi** [comment]
Thanks for the help @Tomcli

Closing this PR since this functionality has been implemented in an iks toolchain: https://github.com/kubeflow/kfp-tekton/pull/435 and https://github.com/kubeflow/kfp-tekton/pull/450.

**Pull request closed without merging**